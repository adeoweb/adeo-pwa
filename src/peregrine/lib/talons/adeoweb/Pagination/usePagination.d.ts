import { PAGER_TILE_TYPE_ELLIPSIS } from 'src/lib/constants/pagination';

type TUsePaginationProps = {
    currentPage: number;
    setPage: (page: number) => void;
    totalPages: number;
};

export type TUsePagination = {
    handleNavBack: () => void;
    handleNavForward: () => void;
    isActiveLeft: boolean;
    isActiveRight: boolean;
    navBackUrl: string;
    navForwardUrl: string;
    tiles: (number | typeof PAGER_TILE_TYPE_ELLIPSIS)[];
    getPageUrl: (page: number) => string;
};

export function usePagination(props: TUsePaginationProps): TUsePagination;
