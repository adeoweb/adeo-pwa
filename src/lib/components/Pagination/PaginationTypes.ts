export interface IPaginationProps {
    currentPage: number;
    setPage: (page: number) => void;
    totalPages: number;
}
