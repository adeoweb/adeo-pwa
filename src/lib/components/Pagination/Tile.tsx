import React, { useCallback, useMemo, FunctionComponent } from 'react';

interface ITileProps {
    isActive: boolean;
    number: number;
    onClick: (page: number) => void;
    getPageUrl: (page: number) => string;
}

const Tile: FunctionComponent<ITileProps> = props => {
    const { isActive, number, onClick, getPageUrl } = props;

    const handleClick = useCallback(
        event => {
            event.preventDefault();
            onClick(number);
        },
        [onClick, number]
    );

    const pageUrl = useMemo(() => getPageUrl(number), [getPageUrl, number]);

    return (
        <li className={`page-item ${isActive ? 'active' : ''}`}>
            <a className="page-link" href={pageUrl} onClick={handleClick}>
                {number}
                {isActive && <span className="sr-only">(current)</span>}
            </a>
        </li>
    );
};

export default Tile;
