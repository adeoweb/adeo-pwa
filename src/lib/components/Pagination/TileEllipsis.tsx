import React, { FunctionComponent } from 'react';

const TileEllipsis: FunctionComponent = () => (
    <li className="page-item">
        <span>...</span>
    </li>
);

export default TileEllipsis;
