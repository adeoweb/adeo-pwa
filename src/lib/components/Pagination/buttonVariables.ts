interface INavIcons {
    [index: string]: string;
}

interface IPage {
    name: string;
    buttonLabel: string;
}

interface INavButtons {
    prevPage: IPage;
    nextPage: IPage;
}

export const navIcons: INavIcons = {
    ChevronLeft: 'icon-angle-left',
    ChevronRight: 'icon-angle-right'
};

export const navButtons: INavButtons = {
    prevPage: {
        name: 'ChevronLeft',
        buttonLabel: 'move to the previous page'
    },
    nextPage: {
        name: 'ChevronRight',
        buttonLabel: 'move to the next page'
    }
};
