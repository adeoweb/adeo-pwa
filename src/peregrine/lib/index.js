// components
export { default as PeregrineContextProvider } from './PeregrineContextProvider';
