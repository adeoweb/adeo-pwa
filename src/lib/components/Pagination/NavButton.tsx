import React, { FunctionComponent, useCallback } from 'react';
import { TFuncKey, useTranslation } from 'react-i18next';

import { navIcons } from './buttonVariables';

interface INavButtonProps {
    name: string;
    label: TFuncKey<'navigation'>;
    active: boolean;
    onClick: () => void;
    url: string;
}

const NavButton: FunctionComponent<INavButtonProps> = ({
    name,
    label,
    active,
    onClick,
    url
}) => {
    const { t } = useTranslation('navigation');

    const handleClick = useCallback(
        event => {
            event.preventDefault();
            onClick();
        },
        [onClick]
    );

    return (
        <li
            className={`page-item ${active ? '' : 'disabled'}`}
            aria-label={t(label)}
        >
            <a
                className="page-link page-link-btn"
                onClick={handleClick}
                href={url}
            >
                <i className={navIcons[name]} />
            </a>
        </li>
    );
};

export default NavButton;
