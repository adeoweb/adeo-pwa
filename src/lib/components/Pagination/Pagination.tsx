import React, { useMemo, FunctionComponent } from 'react';
import { usePagination } from 'src/peregrine/lib/talons/adeoweb/Pagination/usePagination';
import Tile from './Tile';
import TileEllipsis from './TileEllipsis';
import NavButton from './NavButton';
import { navButtons } from './buttonVariables';
import { PAGER_TILE_TYPE_ELLIPSIS } from '../../constants/pagination';
import { IPaginationProps } from 'src/lib/components/Pagination/PaginationTypes';

const Pagination: FunctionComponent<IPaginationProps> = ({
    currentPage,
    setPage,
    totalPages
}) => {
    const {
        handleNavBack,
        handleNavForward,
        isActiveLeft,
        isActiveRight,
        tiles,
        getPageUrl,
        navBackUrl,
        navForwardUrl
    } = usePagination({
        currentPage,
        setPage,
        totalPages
    });

    const navigationTiles = useMemo(
        () =>
            tiles.map((tileNumber, index) =>
                tileNumber === PAGER_TILE_TYPE_ELLIPSIS ? (
                    <TileEllipsis key={`ellipsis-${index}`} />
                ) : (
                    <Tile
                        isActive={tileNumber === currentPage}
                        key={`tile-${tileNumber}`}
                        number={tileNumber}
                        onClick={setPage}
                        getPageUrl={getPageUrl}
                    />
                )
            ),
        [currentPage, tiles, setPage, getPageUrl]
    );

    if (totalPages === 1) {
        return null;
    }

    return (
        <div className="pagination">
            <NavButton
                name={navButtons.prevPage.name}
                label={navButtons.prevPage.buttonLabel}
                active={isActiveLeft}
                onClick={handleNavBack}
                url={navBackUrl}
            />
            {navigationTiles}
            <NavButton
                name={navButtons.nextPage.name}
                label={navButtons.nextPage.buttonLabel}
                active={isActiveRight}
                onClick={handleNavForward}
                url={navForwardUrl}
            />
        </div>
    );
};

export default Pagination;
