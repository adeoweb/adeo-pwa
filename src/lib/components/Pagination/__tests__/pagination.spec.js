import React from 'react';

import { createTestInstance } from '@magento/peregrine';

import Pagination from '../Pagination';

jest.mock('../NavButton', () => () => <i />);
jest.mock('../Tile', () => () => <i />);

jest.mock('@magento/peregrine/lib/talons/Pagination/usePagination', () => {
    const handleLeftSkip = jest.fn();
    const handleRightSkip = jest.fn();
    const handleNavBack = jest.fn();
    const handleNavForward = jest.fn();
    const navBackUrl = jest.fn();
    const navForwardUrl = jest.fn();
    const getPageUrl = '';
    const isActiveLeft = false;
    const isActiveRight = false;
    const tiles = [1, 2];

    const usePagination = jest.fn(() => ({
        handleLeftSkip,
        handleRightSkip,
        handleNavBack,
        handleNavForward,
        navBackUrl,
        navForwardUrl,
        getPageUrl,
        isActiveLeft,
        isActiveRight,
        tiles
    }));

    return { usePagination };
});

const pageControl = {
    currentPage: 1,
    setPage: jest.fn(),
    totalPages: 2
};

test('renders correctly', () => {
    const component = createTestInstance(<Pagination {...pageControl} />);

    expect(component.toJSON()).toMatchSnapshot();
});

test('renders nothing if totalPages is 1', () => {
    const onePage = { ...pageControl, totalPages: 1 };
    const component = createTestInstance(<Pagination {...onePage} />);

    expect(component.toJSON()).toMatchSnapshot();
});
