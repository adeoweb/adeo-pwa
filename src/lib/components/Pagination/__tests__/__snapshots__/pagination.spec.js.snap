// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="pagination"
>
  <i />
  <i />
  <i />
  <i />
</div>
`;

exports[`renders nothing if totalPages is 1 1`] = `null`;
